### PR TITLE
TimeLock+K8s Part 1: Live Reloading HTTP Targets

### PR DESCRIPTION
**Pull request by jeremyk-91**
- [ ] handle 0 server case properly

**Goals (and why)**: Support live reloading of URIs in AtlasDbHttpClients. Eventually we'll wire up the timelock block to be live reloadable, but not in this PR.

**Implementation Description (bullets)**:
- Allow users to provide suppliers for the remote URIs.
- Deprecate non-live-reloading factory methods

**Concerns (what feedback would you like?)**:
- Should I have eliminated the factory methods completely?
- Is there anything better we can do about the log message in `checkAndHandleFailure`?
- Are there any other pitfalls? (Notice there's one in `FailoverFeignTarget:211`)

**Where should we start reviewing?**: Probably `AtlasDbHttpClientsTest` gives a good overview of what this change achieves. Then, `FailoverFeignTarget`

**Priority (whenever / two weeks / yesterday)**: whenever

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2570)

### CODE REVIEW

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java [hunk 1]
```diff
@@ -98,7 +103,7 @@ public void continueOrPropagate(RetryableException ex) {
         checkAndHandleFailure(ex);
         if (retryBehaviour.shouldBackoffAndTryOtherNodes()) {
             int numFailovers = failoverCount.get();
-            if (numFailovers > 0 && numFailovers % servers.size() == 0) {
+            if (numFailovers > 0 && numFailovers % serversListSupplier.get().size() == 0) {
```
Comment (jeremyk-91) [inline atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: possible divide by 0 here

File: atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java [hunk 2]
```diff
@@ -196,14 +201,18 @@ public Retryer clone() {
...
     @Override
     public String url() {
         int indexToHit = failoverCount.get();
         mostRecentServerIndex.set(indexToHit);
-        return servers.get(indexToHit % servers.size());
+        // Need to materialise a value first, otherwise a poorly timed change could result in an out-of-bounds access
+        // (e.g. list has size 5, indexToHit is 4, but in between parameter evaluation and getting the list it reduces
+        // in size).
+        ImmutableList<String> serversList = serversListSupplier.get();
+        return serversList.get(indexToHit % serversList.size());
```
Comment (jeremyk-91) [inline atlasdb-feign/src/main/java/com/palantir/atlasdb/http/FailoverFeignTarget.java]: should also check for divide by 0

**Comment by nziebart** [comment]
It's probably worth thinking about moving to remoting3 before doing any heavy re-working of the http clients.

Once we fix https://github.com/palantir/http-remoting/issues/585, I think we can actually migrate without any issues. Currently we return 503 for not current leader, and 429 for too many blocking requests. Both of these status codes are handled appropriately by remoting3.

**Comment by jeremyk-91** [comment]
Hmm, discussed offline last week. I wasn't aware there was work on the remoting side to deal with this. Will need to confirm the behaviour of remoting clients is right - have a concern about Blocking Timeout 503s, but I guess those will end up getting retried and find their way back to the leader soon.

A bit further out and not the scope of this: at some point if we do #1414 phase 2, we might want to optionally return 308s / 503s depending on whether we know the leader or not. Main concern with random retry is that it is conceivable that we may have a cluster of 40+ nodes with only one leader for a given request, and if you retry uniform randomly there's a p99 of making 181 attempts* before you find it!

*geometric distribution, prob success = 0.025 (40 nodes)

**Comment by nziebart** [comment]
Yeah the blocking timeout 503s won't work optimally with remoting, however given that most clients are now using async timelock, I think it'd be better to focus on continuing to move in that direction.

Blocking timeout 503s also have the issue of not actually timing out if the specified timeout is greater than the blocking timeout.

+1 to returning 308 rather than 503 if we know the leader.

**Comment by jeremyk-91** [comment]
I've gotten https://github.com/palantir/http-remoting/pull/599 out for http-remoting, but propose pushing ahead with this for now given priority of unblocking k8s work and additional testing footprint and changes required as part of remoting.

**Comment by jeremyk-91** [comment]
I've decided to pursue a different approach here that would be much less invasive and not involve messing with the guts of the `FailoverFeignTarget`. New PR incoming soon.

**Pull request closed without merging**